tmail-backend/listeners/src/main/java/com/linagora/tmail/event/CombinedOpenSearchIndexingAndComputeMessageFastViewProjectionListener.java

```diff
@@ -0,0 +1,365 @@
+ */
+
+package com.linagora.tmail.event;
+
+import static org.apache.james.util.ReactorUtils.DEFAULT_CONCURRENCY;
+
+import java.io.IOException;
+import java.util.List;
+
+import jakarta.inject.Inject;
+import jakarta.inject.Named;
+
+import org.apache.commons.collections4.CollectionUtils;
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.backends.opensearch.DocumentId;
+import org.apache.james.backends.opensearch.OpenSearchIndexer;
+import org.apache.james.backends.opensearch.RoutingKey;
+import org.apache.james.events.Event;
+import org.apache.james.events.EventListener;
+import org.apache.james.events.Group;
+import org.apache.james.jmap.api.projections.MessageFastViewPrecomputedProperties;
+import org.apache.james.jmap.api.projections.MessageFastViewProjection;
+import org.apache.james.jmap.event.ComputeMessageFastViewProjectionListener;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageIdManager;
+import org.apache.james.mailbox.MessageUid;
+import org.apache.james.mailbox.SessionProvider;
+import org.apache.james.mailbox.events.MailboxEvents;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.exception.MailboxNotFoundException;
+import org.apache.james.mailbox.model.FetchGroup;
+import org.apache.james.mailbox.model.Mailbox;
+import org.apache.james.mailbox.model.MailboxId;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.model.MessageMetaData;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.model.MessageResult;
+import org.apache.james.mailbox.opensearch.IndexBody;
+import org.apache.james.mailbox.opensearch.MailboxOpenSearchConstants;
+import org.apache.james.mailbox.opensearch.OpenSearchMailboxConfiguration;
+import org.apache.james.mailbox.opensearch.events.OpenSearchListeningMessageSearchIndex;
+import org.apache.james.mailbox.opensearch.json.MessageToOpenSearchJson;
+import org.apache.james.mailbox.store.MailboxSessionMapperFactory;
+import org.apache.james.mailbox.store.mail.MessageMapper;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.metrics.api.MetricFactory;
+import org.opensearch.client.opensearch.core.GetResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.github.fge.lambdas.Throwing;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class CombinedOpenSearchIndexingAndComputeMessageFastViewProjectionListener implements EventListener.ReactiveGroupEventListener {
```

Comment:
I feel ashamed there is so much code duplication.

Reply:
I thought about other alternatives:
  - Continue the bundle listener way: push this topic to James so we have a better reason to refactor those 2 listeners so in the end we can better composite the bundle listener.
  - Use local cache for blob store (file system / memory) -> Can reduce network traffic, but not good in a distributed env where cache eviction would be complicated.
  - Use Distributed cache (e.g. `CachedBlobStore` which relies on Cassandra) -> already enabled for CNB and MU => limited use though: not enable for big attachments/messages, and not solve the network traffic issue anyway.
  
  Any better idea is welcome.

Reply:
Thinking more about this combined listener, it may be a bit complicated when doing Guice binding (depending on both JMAP and OpenSearch enabled). My feeling is we would lose modular implementing this idea.

Reply:
Yet I still want this.

Can we please go to the design board and try to be clever? Like "Add an interface"